### PR TITLE
HPCC-31263 Document security manager user authentication

### PR DESCRIPTION
**Pull request by kenrowland**
Added document covering user authentication for defined security managers

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexinexirisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/SecurityUserAuthentication.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+### LDAP 
+The LDAP security manager uses the configured Active Directory to authenticate users. Once authenticated,
+the user is added to the permissions cache, if enabled, to prevent repeated trips to the AD whenever an
+authentication check is required. 
+
+During an authentication, if enabled, the permission cached is interrogated to see if the user is currently
+cached. If so, the authentication status of the user from the cache is returned. If caching is not enabled,
+digital signatures, if configured for the platform, are used to prevent multiple trips to the AD for each
+user. The signature is used to detect if there has been a change to the user. If changed, the user is marked
```
Comment (afishbeckln) [inline devdoc/SecurityUserAuthentication.md:29]: What kind of change to the user does the signature checking detect?

**Review by afishbeckln — CHANGES_REQUESTED** [review]
One minor comment and one request for clarification.

**Review by afishbeckln — APPROVED** [review]

File: devdoc/SecurityUserAuthentication.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+Since resource access authorization requires an authenticated user, the authorization process also authenticates
+the user before checking authorization. There are a couple of advantages to this 
+* Two separate calls are not required to check authorization (one to verify the user and one to 
+check authorization)
+* The caller can perform third party authorization for specific user access
+
+The _authenticate_ method, or any of its overloads or derivatives, accepts a resource or resource list and a user. 
+These methods authenticate the user first before checking access to the specified resource.
+
+ECLWatch uses user authentication during authorization during its log in process. Instead of first authenticating
```
Comment (JamesDeFabia) [inline devdoc/SecurityUserAuthentication.md:65]: ECL Watch s/b two words

File: devdoc/SecurityUserAuthentication.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+* Two separate calls are not required to check authorization (one to verify the user and one to 
+check authorization)
+* The caller can perform third party authorization for specific user access
+
+The _authenticate_ method, or any of its overloads or derivatives, accepts a resource or resource list and a user. 
+These methods authenticate the user first before checking access to the specified resource.
+
+ECLWatch uses user authentication during authorization during its log in process. Instead of first authenticating 
+the user, it calls an authenticate method passing both the user and the necessary resources for which the user must 
+have access in order to log into ECLWatch.
```
Comment (JamesDeFabia) [inline devdoc/SecurityUserAuthentication.md:67]: ECL Watch s/b two words

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
minor comments inline

**Review by JamesDeFabia — APPROVED** [review]
Looks good from my POV

**Comment by kenrowland** [comment]
@ghalliday Please merge

**Pull request merged**